### PR TITLE
Multigrid hierarchical volume attention for vol_p val/test gap

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Volume pressure in CFD is governed by an elliptic PDE (Laplace/Poisson). The biggest unexplained anomaly in the yi SOTA is the **volume_pressure val/test gap: 4.31% val vs 11.46% test** — a 2.66× split suggesting the model's single-resolution attention cannot resolve the long-range pressure correlations needed to generalize to unseen test geometries.

Multigrid methods (the canonical solver for elliptic PDEs) use coarse-to-fine resolution hierarchy: a coarse grid removes low-frequency errors cheaply, then a fine grid resolves high-frequency residuals. Adding an intra-volume hierarchical attention block — coarse global context via self-attention on a 10% point subset, then cross-attention from full-resolution points to the coarse tokens — should propagate global pressure gradients more efficiently and improve test generalization.

**Distinct from active work:**
- Emma's #654 (dual-tower): *cross-modal* surface↔volume. This is *intra-volume* hierarchical resolution — orthogonal mechanism, can compose later.
- Kohaku's #719 (SDF-stratified norm): fixes the gap via normalization. Multigrid fixes it via receptive field structure — orthogonal.

**Mathematical grounding:** Coarse attention (N_coarse = 6553, 10% of 65536) captures global elliptic pressure field structure (wake, stagnation zone, far-field). Fine cross-attention (N_fine = 65536) broadcasts that context to all volume points — exactly the "prolongation" step of multigrid. The coarse-fine separation decouples low-frequency global from high-frequency local modes.

## Implementation

Add a multigrid block in the **volume branch only**, inserted AFTER Transolver layers and BEFORE the final prediction MLP. No surface changes.

### Step 1 — CLI flags

```python
parser.add_argument('--volume-multigrid-coarse-ratio', type=float, default=0.0,
                    help='Fraction of volume points for coarse global context (0=disabled, e.g. 0.1)')
parser.add_argument('--volume-multigrid-attn-layers', type=int, default=2,
                    help='Number of self-attention layers on coarse subset')
parser.add_argument('--volume-multigrid-cross-heads', type=int, default=8,
                    help='Heads in fine→coarse cross-attention')
```

### Step 2 — Coarse subset selection

```python
def select_coarse(volume_feats, ratio, training: bool):
    N = volume_feats.size(1)
    K = max(64, int(N * ratio))
    if training:
        idx = torch.randperm(N, device=volume_feats.device)[:K]
    else:
        # Deterministic stride for consistent val/test evaluation
        idx = torch.linspace(0, N-1, K, device=volume_feats.device).long()
    return volume_feats[:, idx], idx
```

### Step 3 — Multigrid block

In `build_model` or the model class, when `volume_multigrid_coarse_ratio > 0`, initialize:

```python
self.coarse_attn_layers = nn.ModuleList([
    nn.TransformerEncoderLayer(d_model=n_hidden, nhead=n_heads,
                                dim_feedforward=4*n_hidden, batch_first=True,
                                norm_first=True)
    for _ in range(args.volume_multigrid_attn_layers)
])
self.cross_attn = nn.MultiheadAttention(
    embed_dim=n_hidden,
    num_heads=args.volume_multigrid_cross_heads,
    batch_first=True
)
```

Insert after Transolver volume encode, before prediction MLP:

```python
if self.volume_multigrid_coarse_ratio > 0:
    coarse, _ = select_coarse(volume_features, self.volume_multigrid_coarse_ratio, self.training)
    for layer in self.coarse_attn_layers:
        coarse = layer(coarse)
    attended, _ = self.cross_attn(query=volume_features, key=coarse, value=coarse)
    volume_features = volume_features + attended
```

### Step 4 — Param budget

- Coarse self-attn (2 layers, hidden 512, heads 8): ~3.2M extra params
- Cross-attn (512d, 8h): ~1.05M extra params  
- Total: ~4.3M extra (33% over 12.7M base)

If OOM at bs=4 → drop `--volume-multigrid-coarse-ratio` to 0.05 first (3277 points) before reducing batch size.

## Reproduce command — Arm A (cold-start, full yi SOTA stack + multigrid)

```bash
cd /workspace/senpai/target
torchrun --standalone --nproc_per_node=4 train.py \
  --agent violet \
  --wandb-group yi-round41-multigrid-volume \
  --wandb-name violet/multigrid-coarse010-fullstack \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 \
  --surface-curvature-features k1_k2 \
  --beta-nll-beta 0.5 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --volume-multigrid-coarse-ratio 0.10 \
  --volume-multigrid-attn-layers 2 \
  --volume-multigrid-cross-heads 8 \
  --ema-decay 0.999 --lr-warmup-epochs 1 \
  --epochs 8
```

## Baseline — PR #681 nezuko, W&B run `dc031qpt`

| Metric | yi SOTA val | yi SOTA test | AB-UPT target |
|---|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **7.3767%** | **8.7015%** | — |
| `surface_pressure_rel_l2_pct` | 4.8515% | 4.6236% | 3.82% |
| `wall_shear_rel_l2_pct` | 8.3016% | 8.3214% | 7.29% |
| `volume_pressure_rel_l2_pct` | **4.3066%** | **11.3738%** | 6.08% |
| `wall_shear_x_rel_l2_pct` | 7.1045% | 7.1753% | 5.35% |
| `wall_shear_y_rel_l2_pct` | 9.5832% | 9.5964% | 3.65% |
| `wall_shear_z_rel_l2_pct` | 11.0377% | 10.7383% | 3.63% |

**Primary merge criterion:** `val_primary/abupt_axis_mean_rel_l2_pct` < 7.3767%.

**Secondary diagnostic (novel for this experiment):** `test_primary/volume_pressure_rel_l2_pct` < 11.37% (target ≤9.0% for a strong result). This is the main mechanistic signal — even if val_abupt doesn't beat baseline, a large test_vp improvement would justify a continuation run.

## EP gates

- **EP1 gate**: val_abupt ≤ 20% → continue (cold-start + extra params = slower early convergence)
- **EP3 gate**: val_abupt ≤ 10.5% AND val_vp ≤ 5.5% → continue to full run
- **Terminal**: If val_vp shows ≥15% relative improvement vs baseline (4.31% → ≤3.66%) even without beating val_abupt bar, post results and flag as PROMISING for longer run

## Arm B (conditional on Arm A EP3)

If Arm A passes EP3, queue Arm B at `--volume-multigrid-coarse-ratio 0.05` (half the coarse tokens) to test robustness. Do NOT run Arms A and B in parallel.

## Result reporting

Report per-channel breakdown at EP1, EP3, and terminal. Include param count and VRAM. Headline metric is `test_primary/volume_pressure_rel_l2_pct` — report it prominently.

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"abupt_axis_mean_rel_l2_pct","value":<val_abupt>},"test_metric":{"name":"abupt_axis_mean_rel_l2_pct","value":<test_abupt>}}
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT violet:

## Deviation note: switching to warm-start variant

Cold-start (Arm A as written) doesn't fit our 6h walltime: 5442 steps/epoch × 1.71 s/step + validation ≈ 2.9h/epoch, so we'd only complete ~1.7 epochs. Cold-start of the full SOTA stack typically needs 5-9 epochs (PR #311 was 9, PR #576 was 9), so 1.7 epochs gives a very weak read on the multigrid mechanism.

**Switching strategy:** resume from PR #681 SOTA checkpoint `dc031qpt` (val_abupt 7.3767%, val_vp **4.3066%**) and add the multigrid block on top with fresh zero-initialized residual (= identity at step 0). This:

- Fits ~2 epochs in the budget
- Tests the mechanism cleanly: does multigrid help on top of converged SOTA?
- The headline metric `test_primary/volume_pressure_rel_l2_pct` is directly comparable to baseline
- Per BASELINE.md note, the dc031qpt checkpoint has `surface_input_dim=4` (no curvature/β-NLL/grad-EMA), so I'm omitting `--surface-curvature-features k1_k2`, `--beta-nll-beta`, and `--grad-ema-alpha` for shape compatibility

**Code change:** added non-strict resume that tolerates missing `volume_multigrid.*` keys when `--volume-multigrid-coarse-ratio > 0`, but rejects any other missing or unexpected keys. The output projection of the multigrid cross-attention is zero-initialized so the wrapped model is exactly identity at step 0.

**Launch:**
```
torchrun --standalone --nproc_per_node=4 train.py \
  --resume-from /workspace/senpai/target/artifacts/dc031qpt/checkpoint.pt \
  --agent violet \
  --wandb-group yi-round41-multigrid-volume \
  --wandb-name violet/multigrid-warm-coarse010 \
  --learnable-pe \
  --optimizer lion --lr 1e-5 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --lr-warmup-steps 500 \
  --ema-decay 0.999 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --volume-multigrid-coarse-ratio 0.10 \
  --volume-multigrid-attn-layers 2 \
  --volume-multigrid-cross-heads 8 \
  --epochs 2 \
  --kill-thresholds "1500:val_primary/abupt_axis_mean_rel_l2_pct<10"
```

**Mechanistic acceptance bar (warm-start, single epoch):**
- val_vp baseline (dc031qpt) = 4.3066%; need val_vp ≤ 4.10% (≥5% relative improvement) to call it PROMISING
- val_abupt baseline = 7.3767%; need val_abupt < 7.35% to call it competitive

If either signal is positive, queue Arm B at coarse=0.05. If neither, this is a null result for the mechanism on top of SOTA. Will report results and ask for guidance on whether to attempt cold-start anyway.

**Comment by morganmcg1** [comment]
STUDENT violet:

## Results — Arm A (warm-start, coarse=0.10)

W&B run: [`1udwx3er`](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml/runs/1udwx3er) (`violet/multigrid-warm-coarse010`)
Status: **complete** (2 epochs / 9437 steps, 4.9 h walltime)
Params: **20.06M** (12.7M base + ~7.4M multigrid block)
Peak VRAM: **75.1 GB / GPU** (4×A100, bs=4, 65k surf + 65k vol points)

### Key metrics vs PR #681 baseline (`dc031qpt`)

| Metric | Baseline val | Arm A val | Δ rel | Baseline test | Arm A test | Δ rel |
|---|---:|---:|---:|---:|---:|---:|
| **abupt_axis_mean_rel_l2_pct** | 7.3767 | **7.3266** | **−0.68%** | 8.7015 | **8.6516** | −0.57% |
| surface_pressure_rel_l2_pct | 4.8515 | 4.8180 | −0.69% | 4.6236 | 4.5865 | −0.80% |
| wall_shear_rel_l2_pct | 8.3016 | 8.2577 | −0.53% | 8.3214 | 8.2639 | −0.69% |
| **volume_pressure_rel_l2_pct** | **4.3066** | **4.2542** | **−1.22%** | **11.3738** | **11.3521** | **−0.19%** |
| wall_shear_x_rel_l2_pct | 7.1045 | 7.0805 | −0.34% | 7.1753 | 7.1258 | −0.69% |
| wall_shear_y_rel_l2_pct | 9.5832 | 9.5035 | −0.83% | 9.5964 | 9.5185 | −0.81% |
| wall_shear_z_rel_l2_pct | 11.0377 | 10.9766 | −0.55% | 10.7383 | 10.6752 | −0.59% |

### Acceptance bars (from my deviation note)

- ❌ **PROMISING** (mechanism): val_vp ≤ 4.10% (≥5% rel improvement) — **MISSED** (4.2542%, only −1.22% rel)
- ✅ **COMPETITIVE** (overall): val_abupt < 7.35% — **met** by 0.024pp (7.3266%)
- ❌ **Headline mechanistic signal**: test_vp gap fix — **refuted** (test_vp 11.37→11.35, −0.19% rel; val/test gap 2.66× → 2.67×, **unchanged**)

### What happened — honest analysis

The hypothesis was that intra-volume coarse-fine attention would propagate global elliptic pressure gradients and specifically narrow the val/test volume_pressure gap. **The mechanism's primary claim is essentially refuted**: test_vp moved 0.02pp on a baseline of 11.37%, and the val/test ratio is unchanged.

The val_abupt improvement (−0.05pp) is real but **suspiciously uniform across all channels** (−0.5% to −0.8% rel everywhere). This is the signature of generic fine-tuning gain (2 extra epochs at lr=1e-5 of an already-converged checkpoint), not a localized multigrid effect. A true multigrid mechanism should disproportionately help volume_pressure, but vol_p (−1.22% rel val, −0.19% rel test) is roughly the same magnitude as everything else.

The cross-attn output projection was zero-initialized (identity at step 0) and the multigrid block has 7.4M trainable params, so the block did learn something — but what it learned didn't deliver the hypothesized test-side generalization improvement.

### Why I'm NOT running Arm B

My plan said "if either signal is positive, run Arm B at coarse=0.05". The val_abupt signal is technically positive but I think the time cost (~5h) is hard to justify because:

1. **Halving coarse tokens cannot rescue a refuted mechanism.** The hypothesis says *more* coarse tokens = better global elliptic context. If coarse=0.10 produced no test_vp gap fix, coarse=0.05 has even less coarse capacity and is unlikely to do better on the headline claim.
2. **The proper control isn't Arm B — it's a no-multigrid fine-tune.** The right experiment to attribute the val_abupt gain is to resume `dc031qpt` for 2 epochs at lr=1e-5 *without* the multigrid block. Arm B (smaller coarse ratio) doesn't isolate the mechanism; it just varies a hyperparameter of the already-failed mechanism.
3. **The marginal val gain may already be captured by other landed work.** Per BASELINE.md, single-model SOTA is now 7.3588% (PR #724 norman residual MLP, run `u7obwlh7`). My 7.3266% is only 0.03pp below that.

### Reproduce command

```bash
cd /workspace/senpai/target
torchrun --standalone --nproc_per_node=4 train.py \
  --resume-from /workspace/senpai/target/artifacts/dc031qpt/checkpoint.pt \
  --agent violet \
  --wandb-group yi-round41-multigrid-volume \
  --wandb-name violet/multigrid-warm-coarse010 \
  --learnable-pe \
  --optimizer lion --lr 1e-5 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --lr-warmup-steps 500 \
  --ema-decay 0.999 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --volume-multigrid-coarse-ratio 0.10 \
  --volume-multigrid-attn-layers 2 \
  --volume-multigrid-cross-heads 8 \
  --epochs 2 \
  --kill-thresholds "1500:val_primary/abupt_axis_mean_rel_l2_pct<10"
```

### Suggested follow-ups

1. **No-multigrid fine-tune control** (recommended next step if you want to validate the mechanism). Run dc031qpt for 2 epochs at lr=1e-5 on the same warm-start setup *without* the multigrid block. If that reaches val_abupt ≈ 7.33%, the multigrid block adds nothing. If it stays at 4.30%/7.38%, then the +7.4M params *did* help (just not via the hypothesized mechanism).
2. **Cold-start from-scratch run** (more expensive). The PR's original Arm A: full 8-epoch from-scratch run. The mechanism may need to co-evolve with the rest of the model. ~14h on 4 GPUs — would need 2× walltime budget.
3. **Different attribution: split train/test by geometry distance.** If the val/test gap is driven by *geometry* generalization (test shapes structurally differ from train), no internal architectural fix will close it — better data augmentation or test-time adaptation might.
4. **Park the multigrid block as a no-op flag.** The implementation is clean and zero-init residual, so it could remain in train.py as an off-by-default option for future experiments without polluting the main code path.

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["1udwx3er"],"primary_metric":{"name":"abupt_axis_mean_rel_l2_pct","value":7.32656},"test_metric":{"name":"abupt_axis_mean_rel_l2_pct","value":8.65162}}
```

**Comment by morganmcg1** [comment]
ADVISOR: Merging as yi SOTA.

**val_abupt: 7.3266%** (−0.0322pp vs PR #724 bar of 7.3588%) | **test_abupt: 8.6516%** (−0.0368pp)

Both val and test beat the previous yi SOTA — merging per the merge-on-any-improvement policy.

**Mechanism note:** The primary hypothesis (multigrid architecture reduces vol_p test gap caused by broken restored CFD test cases run_133/158/203/226) is refuted. SDF-stratified analysis showed a uniform ~2.6× val/test gap across all SDF bucket sizes — the multigrid architecture did not selectively reduce the gap for any geometry difficulty tier. The performance improvement is attributed to generic fine-tuning gain at lr=1e-5 warm-started from dc031qpt (PR #681 SOTA), not to the multigrid architecture per se.

The zero-init residual design was correct insurance — no regression at step 0, and the fine-tuning converged to a net improvement. The +7.4M multigrid attention params did not hurt.

**For follow-up:** If multigrid hierarchical attention is worth testing on its own merits (independent of warm-start), it should be tested from a cold start or with a controlled ablation. The vol_p test gap (val ~7.3% vs test ~8.7%) remains an open problem — the 4 broken restored CFD cases in the test set are the likely culprit but require dataset-level investigation.
